src/main.rs

```diff
@@ -149,7 +180,21 @@ impl State {
     ///
     /// No change is made to `self` if an error occurs.
     fn do_move(&mut self, mov: Move) -> Result<NextStep, HanoiError> {
-        unimplemented!()
+        if self.get_tower(mov.from).is_empty() {
+            return Err(HanoiError::EmptyFrom(mov.from))
+        }
+        let disk = self.peek_disk(mov.from);
+        match self.push_disk(mov.to, disk.unwrap()) {
+            Ok(_) => {
+                self.pop_disk(mov.from);
+                if self.done() {
+                    Ok(NextStep::Win)
+                } else {
+                    Ok(NextStep::Continue)
+                }
+             },
+             Err(err) => Err(err)
+        }
```

Comment:
Hmm, if `Ok`, run some code, if `Err` pass the error through. Take a look at `Option::map`!